### PR TITLE
[BEAM-6733] Use Flink 1.6/1.7 prepareSnapshotPreBarrier to replace BufferedOutputManager

### PR DESCRIPTION
**Pull request by mxm**
For Flink version <= 1.5 the Flink Runner has to buffer any elements which are
emitted during a snapshot because the barrier has already been emitted. This
leads to a lot of code complexity.

Flink version >= 1.6 provides a hook to execute an action before the snapshot
barrier is emitted by the operator. We can remove the buffering in favor of
finishing the current bundle in DoFnOperator's prepareSnapshotPreBarrier. The
1.5/1.6/1.7 build setup allows us to make this change with as little code
duplication as possible.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by mxm** [comment]
R @tweise 
CC @aljoscha

File: runners/flink/1.6/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java [hunk 1]
```diff
@@ -0,0 +1,959 @@
...
+      elementCount = 0L;
+      lastFinishBundleTime = getProcessingTimeService().getCurrentProcessingTime();
+    }
+  }
+
+  @Override
+  public void prepareSnapshotPreBarrier(long checkpointId) {
+    // Finish the current bundle before the snapshot barrier is sent downstream
+    // This give us a clean state before taking the actual snapshot
+    invokeFinishBundle();
```
Comment (mxm) [inline runners/flink/1.6/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java]: I should highlight that this is the new method used to finish the bundle before the snapshot barrier is emitted and `snapshot(..)` gets called.

File: runners/flink/1.6/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java [hunk 2]
```diff
@@ -0,0 +1,956 @@
...
+    this.currentOutputWatermark = currentOutputWatermark;
+  }
+
+  /** Factory for creating an {@link FlinkOutputManager} from a Flink {@link Output}. */
+  interface OutputManagerFactory<OutputT> extends Serializable {
+    FlinkOutputManager<OutputT> create(Output<StreamRecord<WindowedValue<OutputT>>> output);
+  }
+
+  /** A {@link DoFnRunners.OutputManager} that forwards data to the Flink runtime. */
+  public static class FlinkOutputManager<OutputT> implements DoFnRunners.OutputManager {
```
Comment (mxm) [inline runners/flink/1.6/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java]: The `OutputManager` could be simplified to remove any buffering.

**Comment by mxm** [comment]
Note that we could get rid of `FlinkSplitStateInternals` and `FlinkSplitStateInternalsTest` entirely for Flink > 1.6.

**Comment by tweise** [comment]
Not sure I like duplicating DoFnOperator and then keeping multiple copies in sync. Can we defer this change until we remove 1.5.x support?

**Comment by mxm** [comment]
There is only one copy. I didn't want to copy DoFnOperator but it proved hard to simplify it without a copy. The good thing is that we have tests in place for 1.5/1.6/1.7 which ensure that nothing breaks across these versions.

In my opinion this change gives us more insensitive to drop 1.5 support. Realistically, it will be hard to remain on multiple Flink versions without code diversion.

**Pull request closed without merging**